### ISSUE TITLE
Show discriminatorValue in discriminatedUnion error message

### ISSUE THREAD

**Issue by deluksic** [colinhacks/zod#3308]
Could we get the [discriminatorValue](https://github.com/colinhacks/zod/blob/master/src/types.ts#L2950) in the error message as well?

Please take a look at the following example:

```typescript
const schema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("apple") }),
  z.object({ type: z.literal("orange") }),
]);

schema.parse({ type: "banana" });
```

Current error message (version 3.22.4):
```
​ Uncaught ZodError: [
  {
    "code": "invalid_union_discriminator",
    "options": [
      "apple",
      "orange"
    ],
    "path": [
      "type"
    ],
    "message": "Invalid discriminator value. Expected 'apple' | 'orange'"
  }
]
```

Expected:
```
...
    "message": "Invalid discriminator value. Expected 'apple' | 'orange', got 'banana'."
...
```

https://playground.solidjs.com/anonymous/230fe5d4-25b0-41dd-b854-8b1b8f99b029
(PS. would love a zod playground, but did not find one suitable)

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
import { zu } from 'zod_utilz'

const schema = z.discriminatedUnion( 'type', [
    z.object( { type: z.literal( 'apple' ) } ),
    z.object( { type: z.literal( 'orange' ) } ),
], {
    errorMap: zu.makeErrorMap( {
        invalid_union_discriminator: ctx =>
            `Invalid discriminator value. Expected 'apple' | 'orange', got '${ ctx.data.type }'.`,
    } ),
} )

const result = schema.safeParse( { type: 'banana' } )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: "invalid_union_discriminator",
//         options: [ "apple", "orange" ],
//         path: [ "type" ],
//         message: "Invalid discriminator value. Expected 'apple' | 'orange', got 'banana'.",
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by deluksic** [comment]
> Is this what you are looking for?

Thanks! That looks very convenient,  I was not aware of `errorMap`. I would still like this to land in Zod itself if possible. If I get a green light I might make a PR as well.

**Comment by JacobWeisenburger** [comment]
you can do this with just zod, but it's much easier with zod_utilz

**Comment by JacobWeisenburger** [comment]
If you want to do this without zod_utilz, here's how:
```ts
const schema = z.discriminatedUnion( 'type', [
    z.object( { type: z.literal( 'apple' ) } ),
    z.object( { type: z.literal( 'orange' ) } ),
], {
    errorMap: ( issue, ctx ) => {
        if ( issue.code === 'invalid_union_discriminator' )
            return { message: `Invalid discriminator value. Expected 'apple' | 'orange', got '${ ctx.data.type }'.` }
        return { message: ctx.defaultError }
    },
} )

const result = schema.safeParse( { type: 'banana' } )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: "invalid_union_discriminator",
//         options: [ "apple", "orange" ],
//         path: [ "type" ],
//         message: "Invalid discriminator value. Expected 'apple' | 'orange', got 'banana'.",
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by deluksic** [comment]
Thank you, I'm using this for now. This issue is however about changing the default behaviour to be more useful.

**Comment by colinhacks** [comment]
This is intention, as it's a general policy in Zod not to leak input data in the default error messages for security reasons. Your custom error map is the recommended solution here. 👍

**Comment by deluksic** [comment]
That makes sense. And would you consider having more detailed error messages in dev vs prod builds?

**Comment by Pnlvfx** [comment]
> This is intention, as it's a general policy in Zod not to leak input data in the default error messages for security reasons. Your custom error map is the recommended solution here. 👍

tbh it only partially make sense,

I came here from joi and I found this library a lot better than it for my use case, 
however I found the validation errors really difficult to read and understand. I think that should be my responsability to not leak the values in production. 

I'm using a library that use zod to validate an api call, there's no way for me to access their code but fork the library and change the error map or all I can do to find the wrong value is guess it.

However still love this library, nothing to say but just a personal thing that is constantly put me in situation where I have to start adding console log and stuff to find a value that could have been provided much more easily by zod itself.
